float64 width
float64 distance
float64 angle